mypy/checkexpr.py

```diff
@@ -3910,6 +3916,15 @@ def visit_conditional_expr(self, e: ConditionalExpr, allow_none_return: bool = F
                 if_type = self.analyze_cond_branch(if_map, e.if_expr, context=else_type,
                                                    allow_none_return=allow_none_return)
 
+        elif if_type == ctx or (if_lit is not None and (if_lit == ctx or if_lit == ctx_lit)):
```

Comment:
Honestly, I don't quite understand why this second call to `analyze_cond_branch` is here at all... It can't think of a case where it is relevant to us the type of the result of the `if` branch to infer the type of the result of the `else` branch. But then again I don't fully understand the semantics of the "type context"...

From what I understand though, it seems silly to pass a literal type as type context to `analyze_cond_branch` and it would probably make more sense to always pass the fallback type instead, which would simplify things back to a simple `if_type == ctx` (after generalizing `if_type` and `ctx` to their respective fallback types if either is a literal).

Maybe someone with a deeper understanding of this logic can chime in? @JukkaL maybe ?

Reply:
Updated with a new approach that feels more theoretically sound and more robust to any change similar to #12395